### DISCUSSION TITLE
Usage without flake-parts

### DISCUSSION THREAD

**Discussion by VanCoding** [Platonic-Systems/process-compose-flake#74] [Q&A]
Hi & thank you very much for this awesome stuff!

I'm currently wondering: Is it possible to use process-compose-flake as a stand-alone library, without flake-parts?

I imagine something like this:

```
let
  processes  = process-compose-flake.lib.buildCommand {
    name = "my-processes";
    imports = [
      inputs.services-flake.processComposeModules.default
    ];
    services.postgres."pg1".enable = true;
  };
in mkShell {
  buildInputs = [
    processes
  ];
}
```

And then being able to
`my-processes up` or `my-processes down`

I know you're probably using the module system provided by flake-parts, and decoupling it would require shipping your own module system. But it would also mean, process-compose-flake could be used within flakes not using flake-parts (flake-utils for example).

**Comment by srid** [comment]
Not sure if that's possible right now. I wouldn't be opposed to accepting a refactor as long as it still keeps things simple. Maybe we can decouple things, so you can use the 'low-level' stuff directly without relying on the high-level flake-parts layer.

Alternatively, maybe there is way for you to use flake-parts just for a subset of your flake (ie., you can use its outputs but without letting it control the flake outputs). cc @roberth

**Reply by VanCoding** [reply]
@srid I've had a look at the source code of flake-parts, process-compose-flake and services-flake and came up with the following:

```
{
    inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    inputs.flake-utils.url = "github:numtide/flake-utils";
    inputs.process-compose-flake.url = "github:Platonic-Systems/process-compose-flake";
    inputs.services-flake.url = "github:juspay/services-flake";
    outputs = {nixpkgs, flake-utils, process-compose-flake, services-flake, self}:  flake-utils.lib.eachDefaultSystem (system: let pkgs = import nixpkgs { inherit system; }; in {
        packages.processes = (nixpkgs.lib.evalModules {
            specialArgs = { inherit pkgs; name = "processes"; };
            modules = [
                "${process-compose-flake}/nix/process-compose"
                services-flake.processComposeModules.default
                ({pkgs,lib,...}:{
                    services.postgres."pg1".enable = true;
                })
            ];
        }).config.outputs.package;
    });
}
```

Then running `nix run .#processes up` actually starts postgres, so it's definitely possible to use process-compose-flake without flake parts. So it'd be really cool if there was an official utility function to do so :)

**Comment by adrian-gierakowski** [comment]
Also interested as I currently hacked around it in similar way since I’m not even using flakes at work yet (waiting for lazy trees)
